api/src/main/java/org/openmrs/api/impl/DiagnosisServiceImpl.java

```diff
@@ -0,0 +1,148 @@
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.api.impl;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.openmrs.Diagnosis;
+import org.openmrs.DiagnosisOrder;
+import org.openmrs.Encounter;
+import org.openmrs.Patient;
+import org.openmrs.api.APIException;
+import org.openmrs.api.DiagnosisService;
+import org.openmrs.api.context.Context;
+import org.openmrs.api.db.DiagnosisDAO;
+import org.openmrs.parameter.EncounterSearchCriteria;
+import org.openmrs.parameter.EncounterSearchCriteriaBuilder;
+/**
+ * API functionality for Diagnoses.
+ *
+ * @since 2.2
+ */
+public class DiagnosisServiceImpl extends BaseOpenmrsService implements DiagnosisService {
+
+	private DiagnosisDAO dao;
+
+	public DiagnosisDAO getDao() {
+		return dao;
+	}
+
+	public void setDao(DiagnosisDAO dao) {
+		this.dao = dao;
+	}
+
+	/**
+	 * @see org.openmrs.api.DiagnosisService#save(Diagnosis)
+	 */
+	public Diagnosis save(Diagnosis diagnosis) {
+		return dao.saveDiagnosis(diagnosis);
+	}
+
+	/**
+	 * @see org.openmrs.api.DiagnosisService#voidDiagnosis(Diagnosis, String)
+	 */
+	public Diagnosis voidDiagnosis(Diagnosis diagnosis, String voidReason) {
+		return dao.saveDiagnosis(diagnosis);
+	}
+
+	/**
+	 * @see org.openmrs.api.DiagnosisService#getDiagnosisByUuid(String)
+	 */
+	public Diagnosis getDiagnosisByUuid(String uuid) {
+		return dao.getDiagnosisByUuid(uuid);
+	}
+
+	/**
+	 * @see org.openmrs.api.DiagnosisService#getDiagnoses(Patient, Date)
+	 */
+	public List<Diagnosis> getDiagnoses(Patient patient, Date fromDate) {
+
+		EncounterSearchCriteria encounterSearchCriteria = new EncounterSearchCriteriaBuilder().setPatient(patient)
+			.setFromDate(fromDate).createEncounterSearchCriteria();
+		List<Encounter> encounters = Context.getEncounterService().getEncounters(encounterSearchCriteria);
+
+		List<Diagnosis> diagnoses = new ArrayList<Diagnosis>();
+
+		for (Encounter encounter : encounters) {
+			diagnoses.addAll(encounter.getDiagnoses());
+		}
+		return diagnoses;
+	}
+
+	/**
+	 * @see org.openmrs.api.DiagnosisService#getPrimaryDiagnoses(Encounter)
+	 */
+	public List<Diagnosis> getPrimaryDiagnoses(Encounter encounter) {
+		List<Diagnosis> diagnoses = new ArrayList<Diagnosis>();
+		if (encounter != null && !encounter.getVoided()) {
+
+			for (Diagnosis diagnosis :
+				encounter.getDiagnoses()) {
+				// check if the diagnosis is given the highiest importance. 
+				if (diagnosis.getDiagnosisOrder() == DiagnosisOrder.PRIMARY) {
+					diagnoses.add(diagnosis);
+				}
+			}
+		}
+		return diagnoses;
+	}
+
+	/**
+	 * @see org.openmrs.api.DiagnosisService#hasDiagnosis(Encounter, Diagnosis)
+	 */
+	public boolean hasDiagnosis(Encounter encounter, Diagnosis diagnosis) {
+
+		for (Diagnosis diagnosisFound :
+			encounter.getDiagnoses()) {
```

Comment:
Does `encounter.getDiagnoses()` only return non voided diagnoses? Because I don't think you'd want to return true if the found diagnosis is voided